analytics-ios-mcvid/Classes/SEGMCVIDTracker.m

```diff
@@ -32,10 +37,170 @@ -(id)initWithOrganizationId:(NSString *)organizationId region:(NSString *)region
+    NSString *errorDomain = @"Segment-Adobe";
+    NSString *marketingCloudIdKey = @"d_mid";
+
+    NSString *marketingCloud = nil;
+    NSString *advertisingId = nil;
+
+    NSURL *url = [self createURL:advertisingId organizationId:organizationId marketingCloudId:marketingCloud];
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
+    NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
+    NSURLSession *session = [NSURLSession sessionWithConfiguration:config];
+
+    [[session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        void (^callbackWithCode)(MCVIDAdobeErrorCode code, NSString *message, NSError *error) = ^void(MCVIDAdobeErrorCode code, NSString *message, NSError *error) {
+            MCVIDAdobeError *adobeError = [[MCVIDAdobeError alloc] initWithCode:code message:message error:error];
+            NSError *compositeError = [NSError errorWithDomain:errorDomain code:adobeError.code userInfo:@{MCVIDAdobeErrorKey:adobeError}];
+            completion(nil, compositeError);
+        };
+        
+        NSDictionary *dictionary = nil;
+        @try {
+            dictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
+        } @catch (NSException *exception) {
+            return callbackWithCode(MCVIDAdobeErrorCodeClientSerializationError, @"Deserializing the JSON response failed", nil);
+        }
+        
+        NSString *marketingCloudId = dictionary[marketingCloudIdKey];
+
+        
+        // or { ..., "errors": [{ "code": 2, "msg": "error" } ... ], ... }
+        NSDictionary *errorDictionary = dictionary[errorResponseKey];
+        NSError *errorObject = dictionary[errorResponseKey][0];
+        NSString *errorMessage = dictionary[@"errors"][0][@"msg"];
+        
+        NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*)response;
+        NSInteger responseStatusCode = [httpResponse statusCode];
+        
+        //Logic for exponential backoff algorithm
+        NSUInteger milliSecondsToWait = currentRetryCount * currentRetryCount;
+        NSUInteger milliSecondsWaited = currentRetryCount * (currentRetryCount + 1)  * ((2 * currentRetryCount) + 1)/6;
+
+        if ((currentRetryCount > maxRetryCount) && errorObject) {
+            return callbackWithCode(MCVIDAdobeErrorCodeServerError, errorMessage, errorObject);
+        }
+
+        if ((milliSecondsWaited / 1 >= maxRetryTimeSecs) && errorObject) {
+            return callbackWithCode(MCVIDAdobeErrorCodeServerError, errorMessage, errorObject);
+        }
+
+        if (responseStatusCode == 200 && (!errorObject) ){
+            completion(marketingCloudId, nil);
+        } else {
+            dispatch_time_t delay = dispatch_time(DISPATCH_TIME_NOW, NSEC_PER_SEC * milliSecondsToWait);
+            self.backgroundQueue = dispatch_queue_create("com.exmaple.MyCustomQueue", NULL);
```

Comment:
I think we should create this queue once in initialization, and re-use it (as it can be expensive to create).